camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java

```diff
@@ -0,0 +1,383 @@
+    return readMessages(new Path(destinationPath, topic));
+  }
+
+  private List<MyMessage> readMessages(Path path) throws IOException, InstantiationException, IllegalAccessException {
+    List<MyMessage> messages = new ArrayList<MyMessage>();
+
+    try {
+      for (FileStatus file : fs.listStatus(path)) {
+        if (file.isDir()) {
+          messages.addAll(readMessages(file.getPath()));
+        } else {
+          SequenceFile.Reader reader = new SequenceFile.Reader(fs, file.getPath(), new Configuration());
+          try {
+            LongWritable key = (LongWritable) reader.getKeyClass().newInstance();
+            Text value = (Text) reader.getValueClass().newInstance();
+
+            while (reader.next(key, value)) {
+              messages.add(gson.fromJson(value.toString(), MyMessage.class));
+            }
+          } finally {
+            reader.close();
+          }
+        }
+      }
+    } catch (FileNotFoundException e) {
+      System.out.println("No camus messages were found in [" + path + "]");
+    }
+
+    return messages;
+  }
+
+  private static class TestEtlInputFormat extends EtlInputFormat {
+    public TestEtlInputFormat() {
+      super();
+    }
+
+    public static void setLogger(Logger log) {
+      EtlInputFormat.setLogger(log);
+    }
+    
+    @Override
+    public SimpleConsumer createSimpleConsumer(JobContext context, String host, int port) {
+      return _simpleConsumer;
+    }
+  }
+  
+  private static class MyMessage {
+
+    private int number;
+
+    // Used by Gson
+    public MyMessage() {
+    }
+
+    public MyMessage(int number) {
+      this.number = number;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj instanceof MyMessage))
+        return false;
+
+      MyMessage other = (MyMessage) obj;
+
+      return number == other.number;
+    }
+  }
+
+  private static void resetCamus() throws NoSuchFieldException, IllegalAccessException {
+    _mocks.clear();
+    _simpleConsumer = null;
+    
+    // The EtlMultiOutputFormat has a static private field called committer which is only created if null. The problem is this
+    // writes the Camus metadata meaning the first execution of the camus job defines where all committed output goes causing us
+    // problems if you want to run Camus again using the meta data (i.e. what offsets we processed). Setting it null here forces
+    // it to re-instantiate the object with the appropriate output path
+
+    Field field = EtlMultiOutputFormat.class.getDeclaredField("committer");
+    field.setAccessible(true);
+    field.set(null, null);
+  }
+
+  // Kafka Scala layer only provides desearilization as a way to create response, it's very hairy on 
+  // the exact order of those bytes.
+  private static TopicMetadataResponse createTopicMetadataResponseFromBytes(TopicMetadataRequest request) {      
```

Comment:
But then how are you going to set that boolean argument?  So far the design was open the setupJobX() method for each test case to let each test case to explicitly setup the desired behavior for each mock interaction.

If we want to generalize this approach, we would have to come up with the data structure to represent all test data scenarios.